### PR TITLE
Connection pool

### PR DESCRIPTION
**Pull request by EKrysenko**
Create ConnectionPool class.
Create simple test to check getConnection method.
Add MySQL/jdbc dependency at pom.xml

### CODE REVIEW

File: src/main/java/com/epam/webelecty/persistence/database/ConnectionPool.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    public static void dispose() throws SQLException {
+        if (instance != null) {
+            instance.clearConnectionQueue();
+            instance = null;
+        }
+    }
+
+    public synchronized static ConnectionPool getInstance() {
```
Comment (SMishAlex) [inline src/main/java/com/epam/webelecty/persistence/database/ConnectionPool.java]: use better singleton implementation (only first call should be synchronized)

File: src/main/java/com/epam/webelecty/persistence/database/ConnectionPool.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        }
+        connectionQueue = new ArrayBlockingQueue<>(poolSize);
+        for (int i = 0; i < poolSize; i++) {
+            Connection connection = null;
+            try {
+                connection = DriverManager.getConnection(url, user, password);
+            } catch (SQLException e) {
+                e.printStackTrace();
+            }
+            connectionQueue.offer(connection);
```
Comment (SMishAlex) [inline src/main/java/com/epam/webelecty/persistence/database/ConnectionPool.java]: this implementation could provide NPE on client cod (via getConnection)

File: src/main/java/com/epam/webelecty/persistence/database/ConnectionPool.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            // Connection not added", e
+        }
+    }
+
+    private ConnectionPool() {
+        DBResourceManager dbResourceManager = DBResourceManager.getInstance();
+        this.driverName = dbResourceManager.getValue(DBParameter.DB_DRIVER);
+        this.url = dbResourceManager.getValue(DBParameter.DB_URL);
+        this.user = dbResourceManager.getValue(DBParameter.DB_USER);
+        this.password = dbResourceManager.getValue(DBParameter.DB_PASSWORD);
```
Comment (SMishAlex) [inline src/main/java/com/epam/webelecty/persistence/database/ConnectionPool.java]: we are using spring so this parameters will be injected with configuration, as far as singleton pattern will be realised

File: src/test/com/epam/webelecty/persistence/database/ConnectionPoolTest.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.ResultSet;
+
+public class ConnectionPoolTest {
+
+
+    @Test
+    @SneakyThrows
```
Comment (SMishAlex) [inline src/test/com/epam/webelecty/persistence/database/ConnectionPoolTest.java]: first of all don't separate annotations with its target entity, secondary you dont need to use @SneakyThrows here just add throws Exception to your test signature, and the last but not the least: call test methods the way the reader can understand what they do without reading the body of method

File: src/test/com/epam/webelecty/persistence/database/ConnectionPoolTest.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+    public void getConnection1() {
+        Connection connection = DriverManager.getConnection("jdbc:mysql://sql11.freemysqlhosting.net:3306/sql11226348",
+                "sql11226348", "N7XThNApKx");
```
Comment (SMishAlex) [inline src/test/com/epam/webelecty/persistence/database/ConnectionPoolTest.java]: this parametrs should be getted from resources too

File: src/test/com/epam/webelecty/persistence/database/ConnectionPoolTest.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        ResultSet resultSet = connection.prepareStatement("select * from users").executeQuery();
+        while (resultSet.next()) {
+            System.out.println(resultSet.getString("name"));
+            System.out.println(resultSet.getString("lastName"));
+        }
```
Comment (SMishAlex) [inline src/test/com/epam/webelecty/persistence/database/ConnectionPoolTest.java]: there cant be tests without assertions

**Review by SMishAlex — COMMENTED** [review]
I think we will discuss it after tomorrow's stand up

File: pom.xml [hunk 7]
```diff
@@ -87,7 +87,7 @@
           <version>2.2</version>
           <configuration>
             <server>tomcat-development-server</server>
-            <port>8080</port>
+            <port>8082</port>
```
Comment (SMishAlex) [inline pom.xml:98]: what's wrong with 8080?

File: src/main/java/com/epam/webelecty/controllers/TestController.java [hunk 8]
```diff
@@ -1,8 +1,22 @@
...
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.ResponseBody;
 
+
+@Setter
 @Controller
 public class TestController {
+    private ConnectionPool connectionPool;
+
+    @RequestMapping(value = "/", method = RequestMethod.GET)
```
Comment (SMishAlex) [inline src/main/java/com/epam/webelecty/controllers/TestController.java:16]: you can also use GetMapping annotation (just for information)

File: src/main/java/com/epam/webelecty/persistence/database/ConnectionPool.java [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.SQLException;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+
+@Log4j2
+public class ConnectionPool {
+    private static ConnectionPool instance;
```
Comment (SMishAlex) [inline src/main/java/com/epam/webelecty/persistence/database/ConnectionPool.java]: I guess you don't need it anymore

File: src/main/java/com/epam/webelecty/persistence/database/ConnectionPool.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        try {
+            connection = connectionQueue.take();
+        } catch (InterruptedException e) {
+            log.error("Free connection waiting interrupted. Returned `null` connection", e);
+        }
+        return connection;
+    }
+
+    public void dispose() throws SQLException {
+        System.out.println("pre destroy custom method");
```
Comment (SMishAlex) [inline src/main/java/com/epam/webelecty/persistence/database/ConnectionPool.java]: remove or replace with log

File: src/main/java/com/epam/webelecty/persistence/database/DBParameter.java [hunk 11]
```diff
@@ -0,0 +1,11 @@
+package com.epam.webelecty.persistence.database;
+
+public class DBParameter {
```
Comment (SMishAlex) [inline src/main/java/com/epam/webelecty/persistence/database/DBParameter.java]: I think you dont use this class anymore

File: src/main/java/com/epam/webelecty/persistence/database/DBResourceManager.java [hunk 12]
```diff
@@ -0,0 +1,17 @@
+package com.epam.webelecty.persistence.database;
+
+import java.util.ResourceBundle;
+
+public class DBResourceManager {
```
Comment (SMishAlex) [inline src/main/java/com/epam/webelecty/persistence/database/DBResourceManager.java]: as far as this one

File: src/test/com/epam/webelecty/persistence/database/ConnectionPoolTest.java [hunk 13]
```diff
@@ -0,0 +1,47 @@
+package com.epam.webelecty.persistence.database;
+
+import org.junit.Assert;
+import org.junit.Test;
+import java.sql.Connection;
+import java.sql.SQLException;
+
+public class ConnectionPoolTest {
```
Comment (SMishAlex) [inline src/test/com/epam/webelecty/persistence/database/ConnectionPoolTest.java]: let it be for now

File: pom.xml (hunk 7 above)
Comment (EKrysenko) [reply pom.xml:98]: Didn`t connect. Donno why.

File: src/main/java/com/epam/webelecty/controllers/TestController.java (hunk 8 above)
Comment (EKrysenko) [reply src/main/java/com/epam/webelecty/controllers/TestController.java:16]: That`s from Egor`s commit. Just for testing.

File: src/main/java/com/epam/webelecty/persistence/database/ConnectionPool.java (hunk 10 above)
Comment (EKrysenko) [reply src/main/java/com/epam/webelecty/persistence/database/ConnectionPool.java]: Oh, it`s a freaky trying to figure how it works with annotation.
Used together with custom postConstruct method, forget to remove.
Next time I will not use "sout'

File: src/main/java/com/epam/webelecty/persistence/database/ConnectionPool.java (hunk 9 above)
Comment (EKrysenko) [reply src/main/java/com/epam/webelecty/persistence/database/ConnectionPool.java]: Fixed

**Pull request merged**